lib/seattleflu/id3c/cli/command/offer_uw_testing.py

```diff
@@ -0,0 +1,232 @@
+"""
+Offer UW Husky Coronavirus Testing.
+
+Updates REDCap to trigger offers of testing to the individuals in the priority
+queue based on the testing capacity quotas for the given time period.
+
+This command is idempotent and can be safely re-run to e.g. pickup a missed
+cronjob or troubleshoot or fix bugs.  Running the command more than once won't
+release extra tests above the quota.
+"""
+import click
+import enum
+import json
+import logging
+import os
+from datetime import date
+from more_itertools import bucket
+from id3c.cli import cli
+from id3c.cli.command import with_database_session, DatabaseSessionAction
+from id3c.cli.redcap import Project, InstrumentStatus
+from id3c.db.session import DatabaseSession
+from id3c.json import dump_ndjson
+from ...utils import unwrap
+
+
+LOG = logging.getLogger(__name__)
+
+TODAY = date.today().isoformat()
+
+
+@cli.command("offer-uw-testing", help = __doc__)
+@with_database_session(pass_action = True)
+
+@click.option("--at",
+    metavar = "<timestamp>",
+    default = "now",
+    help    = unwrap("""
+        Use quota for the given <timestamp>.
+
+        With --dry-run, can be used to test a quota schedule.  Otherwise, can
+        be used to apply unused quota from past timespans.  (Though you should
+        do that with caution, especially across day boundaries.)
+
+        Defaults to the current time."""))
+
+@click.option("--log-offers/--no-log-offers",
+    help    = "Write REDCap offers of testing to stdout as an NDJSON stream.",
+    default = False)
+
+def offer_uw_testing(*, at: str, log_offers: bool, db: DatabaseSession, action: DatabaseSessionAction):
+    LOG.debug(f"Offering UW Husky Coronavirus Testing @ {at}")
+
+    dry_run = action is DatabaseSessionAction.DRY_RUN
+
+    # This uses a mutable quota to track available vs. used testing capacity
+    # for given time periods.  An alternate approach would be to use a
+    # log/ledger (like we keep in receiving.* tables) which records credits
+    # (tests capacity scheduled for release at a certain time) and debits
+    # (tests offered at a certain time).  While this requires recalculating the
+    # balance every run, we would be able to query when tests were released and
+    # keep more metadata about that.  These same benefits could be realized by
+    # turning our normal logging output into structured event logs.  I think
+    # that's preferrable, so decided not to implement as a ledger right now.
+    #   -trs, 17 Sept & 13 Oct 2020
+
+    # Lookup the quota for the current time, locking it for update at the end
+    # after we make offers.
+    #
+    # XXX TODO: As a future improvement, automatically pick up any remaining
+    # quota left from _past_ timespans in the current day.
```

Comment:
+1 This would be nice to have operationally

Reply:
Agreed. I left it out because it's a bit nuanced to get right and additional complexity right now that we probably don't need (assuming that our queue is never close to empty and/or our quota for a timespan is never very large).